src/useField.js

```diff
@@ -0,0 +1,205 @@
+// @flow
+import * as React from 'react'
+import { fieldSubscriptionItems } from 'final-form'
+import flattenSubscription from './flattenSubscription'
+import type { FieldSubscription, FieldState, FormApi } from 'final-form'
+import type { UseFieldConfig, FieldInputProps, FieldRenderProps } from './types'
+import isReactNative from './isReactNative'
+import getValue from './getValue'
+import ReactFinalFormContext from './context'
+
+const all: FieldSubscription = fieldSubscriptionItems.reduce((result, key) => {
+  result[key] = true
+  return result
+}, {})
+
+export const defaultFormat = (value: ?any, name: string) =>
+  value === undefined ? '' : value
+export const defaultParse = (value: ?any, name: string) =>
+  value === '' ? undefined : value
+
+const useField = (
+  name: string,
+  {
+    allowNull,
+    component,
+    defaultValue,
+    format,
+    formatOnBlur,
+    initialValue,
+    isEqual,
+    multiple,
+    parse,
+    subscription,
+    type,
+    validate,
+    validateFields,
+    value: _value
+  }: UseFieldConfig = {}
+): FieldRenderProps => {
+  const reactFinalForm: ?FormApi = React.useContext(ReactFinalFormContext)
+  // istanbul ignore next
+  if (process.env.NODE_ENV !== 'production' && !reactFinalForm) {
+    console.error(
+      'Warning: useField must be used inside of a ReactFinalForm component'
+    )
+  }
+
+  // keep ref to most recent copy of validate function
+  const validateRef = React.useRef(validate)
+  validateRef.current = validate
+
+  const register = (callback: FieldState => void) =>
+    reactFinalForm
+      ? reactFinalForm.registerField(name, callback, subscription || all, {
+          defaultValue,
+          getValidator: () => validateRef.current,
+          initialValue,
+          isEqual,
+          validateFields
+        })
+      : () => {}
+
+  const firstRender = React.useRef(true)
+
+  // synchronously register and unregister to query field state for our subscription on first render
+  const [state, setState] = React.useState<FieldState>(
+    (): FieldState => {
+      let initialState: FieldState = {}
+      register(state => {
+        initialState = state
+      })()
+      return initialState
+    }
+  )
+
+  const flattenedSubscription = flattenSubscription(subscription || all)
+  // useDependenciesDebugger({
+  //   name,
+  //   defaultValue,
+  //   validate,
+  //   initialValue,
+  //   isEqual,
+  //   validateFields,
+  //   reactFinalForm
+  // })
+  React.useEffect(
+    () =>
+      register(state => {
+        if (firstRender.current) {
+          firstRender.current = false
```

Comment:
in theory the state might change between 1st render and this effect, you probably should conditionally setState if it has changed in between

Reply:
Not sure I understand...

Reply:
From what I understand `register` gives `state` synchronously upon registration (just like `subscribe`).

So this is a classic hooks-subscription problem. Technically the state can change between your first register in lazy initialization setState and subscription point in `useEffect`.

Reply:
> From what I understand `register` gives `state` synchronously upon registration (just like `subscribe`).

Yes. That's correct.

How do I
> conditionally `setState` if it has changed in between

? I'm not 100% on what you mean by that.

Reply:
if its first render you must compare receives state against the one in react's state and if they differ you should setState (thus triggering "double render", but it is what it is)

Reply:
I've been trying to avoid including `deepEqual` code in the library. Perhaps it's unavoidable?

Reply:
why cant this be based on shallowEqual? which part of the state is not comparable as easily?

Reply:
Same here. Gonna go think about something else and see if any insights arrive. 😄